File: Leetcode/543-Diameter-Of-Binary-Tree/DiameterOfBinaryTree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func diameterOfBinaryTree(root *TreeNode) int {
	res := getDiameter(root)
	return res[0]
}

func getDiameter(node *TreeNode) [2]int {
	res := [2]int{0, 0}
	if node == nil {
		return res
	}
	left := getDiameter(node.Left)
	right := getDiameter(node.Right)
	res[0] = max(left[1]+right[1], max(left[0], right[0]))
	res[1] = max(left[1], right[1]) + 1
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: Leetcode/543-Diameter-Of-Binary-Tree/DiameterOfBinaryTree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class DiameterOfBinaryTree {
  public int diameterOfBinaryTree(TreeNode root) {
        int[] res = getDiameter(root);
        return res[0];
    }

    /**
     * @return [0] for round, [1] for edge.
     */
    private int[] getDiameter(TreeNode node) {
        int[] res = new int[2];
        if (node == null) {
            return res;
        }
        int[] left = getDiameter(node.left);
        int[] right = getDiameter(node.right);
        res[0] = Math.max(left[1]+right[1], Math.max(left[0], right[0]));
        res[1] = Math.max(left[1], right[1])+1;
        return res;
    }
}
